lib/sources.nix

```diff
@@ -101,4 +101,201 @@ rec {
   pathHasContext = builtins.hasContext or (lib.hasPrefix builtins.storeDir);
 
   canCleanSource = src: src ? _isLibCleanSourceWith || !(pathHasContext (toString src));
+
+  # Splits a filesystem path into its components.
+  splitPath = path: lib.splitString "/" (toString path);
+
+  # Turns a list of path components into a tree, e.g.
+  #
+  #     ["a" "b" "c1"]
+  #     ["a" "b" "c2"]
+  #     ["a" "b" "c3"]
+  #     ["a" "x"     ]
+  #
+  # becomes:
+  #
+  #     { a = { b = { c1 = null; c2 = null; c3 = null }; x = null; }; }
+  pathComponentsToTree = paths: with lib;
+    foldl (tree: path: recursiveUpdate tree (setAttrByPath path null)) {} paths;
+
+  # Returns true if and only if any prefix of the given `path` leads to a leaf
+  # (`null`) in the given `tree` (nested attrset).
+  # That is: "If we go down the tree by the given path, do we hit a leaf?"
+  #
+  # Example: For the tree
+  #
+  #     a
+  #       b-leaf
+  #       c-leaf
+  #       d
+  #         e-leaf
+  #     f-leaf
+  #
+  # represented as attrset
+  #
+  #     { a = { b = null; c = null; d = { e = null; }; }; f = null; }
+  #
+  # we have
+  #
+  #     isPrefixOfLeafPath [a] == true
+  #     isPrefixOfLeafPath [x] == false
+  #     isPrefixOfLeafPath [a b] == true
+  #     isPrefixOfLeafPath [a b c] == true
+  #     isPrefixOfLeafPath [a b c x] == true
+  #     isPrefixOfLeafPath [a d] == false
+  isPrefixOfLeafPath = path: tree:
+    if tree == null
+      then true
+      else
+        if path == []
+          then false
+          else
+            let
+              component = builtins.head path;
```

Comment:
All these builtins (except `builtins.trace`) are also reexported by lib, so you can drop `builtins.`

Reply:
That doesn't work, as `with lib;` is only enabled on a per-function level and `isPrefixOfLeafPath` doesn't have it, so it seems that it's needed.

I could add a `with lib;` to it, but is there a drawback/benefit of using `lib` vs `builtins`? It appeared to me that it's nice to only use builtins when possible, so that it's clear that you could easily copy out the function if needed / that it has not dependencies.

Reply:
It's preferable to avoid `with`, so Nix can perform scope checking and improve lookup performance of performance-sensitive code like this.